### PR TITLE
[SPARK-7059][SQL] Create a DataFrame join API to facilitate equijoin.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -343,6 +343,35 @@ class DataFrame private[sql](
   }
 
   /**
+   * Inner equi-join with another [[DataFrame]] using the column.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: nit: the _given_ column. ?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -343,6 +343,35 @@ class DataFrame private[sql](
...
+   * {{{
+   *   // Joining df1 and df2 using the column "user_id"
+   *   df1.join(df2, "user_id")
+   * }}}
+   *
+   * Note that if you perform a self-join using this function, you wouldn't be able to reference
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: ... if you perform a self-join _without aliasing the input DataFrames_ ...
nit: you _will not`_ be able ?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 3]
```diff
@@ -343,6 +343,35 @@ class DataFrame private[sql](
   }
 
   /**
+   * Inner equi-join with another [[DataFrame]] using the column.
+   * {{{
+   *   // Joining df1 and df2 using the column "user_id"
+   *   df1.join(df2, "user_id")
+   * }}}
+   *
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: I would also state that similar to SQL USING the join column will only appear once in the output.

**Comment by rxin** [comment]
Addressed the comments and added Python too.

**Pull request closed without merging**